assets/blueprints/genesis_helper/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+                let initial_stake_amount = genesis_data
+                    .stakes
+                    .get(&validator_idx)
+                    .map(|stakes| {
+                        stakes
+                            .iter()
+                            .map(|(_, xrd_amount)| xrd_amount)
+                            .cloned()
+                            .sum()
+                    })
+                    .unwrap_or(Decimal::ZERO);
+                let initial_stake_bucket = whole_lotta_xrd.take(initial_stake_amount);
+                validators_with_initial_stake.push((
+                    validator.key,
+                    validator.component_address,
+                    initial_stake_bucket,
+                ));
+            }
+
+            let lp_buckets: Vec<Bucket> = Runtime::call_function(
+                EPOCH_MANAGER_PACKAGE,
+                EPOCH_MANAGER_BLUEPRINT,
+                EPOCH_MANAGER_CREATE_IDENT,
+                scrypto_encode(&EpochManagerCreateInput {
+                    validator_owner_token,
+                    component_address: epoch_manager_component_address,
+                    validator_set: validators_with_initial_stake,
+                    initial_epoch,
+                    rounds_per_epoch,
+                    num_unstake_epochs,
+                })
+                .unwrap(),
+            );
+
+            // ...and distribute the LP tokens to stakers
+            for (validator_idx, mut lp_bucket) in lp_buckets.into_iter().enumerate() {
+                let stakes = genesis_data.stakes.remove(&validator_idx).unwrap_or(vec![]);
+                for (account_idx, stake_xrd_amount) in stakes {
+                    // TODO: currently xrd amount matches stake tokens amount, but can this change later on?
+                    let stake_bucket = lp_bucket.take(stake_xrd_amount);
+                    let account_address = genesis_data.accounts[account_idx];
+                    let _: () = Runtime::call_method(
+                        account_address,
+                        "deposit",
+                        scrypto_encode(&AccountDepositInput {
+                            bucket: stake_bucket,
+                        })
+                        .unwrap(),
+                    );
+                }
+                lp_bucket.drop_empty();
+            }
+
+            // Allocate XRD
+            for (account_idx, xrd_amount) in genesis_data.xrd_balances.into_iter() {
+                let account_address = genesis_data.accounts[account_idx];
+                let bucket = whole_lotta_xrd.take(xrd_amount);
+                let _: () = Runtime::call_method(
+                    account_address,
+                    "deposit",
+                    scrypto_encode(&AccountDepositInput { bucket }).unwrap(),
+                );
+            }
+
+            // return the remainder
+            whole_lotta_xrd
+        }
+
+        fn create_resource(
+            bech32_encoder: &Bech32Encoder,
```

Comment:
I don't think we should have the bech32 encoder in here. This should be handled by frontends rather than in the engine since we've agreed in the past that the engine should be abstracted from Bech32 + NetworkDefinition.

Reply:
Okay, sure :+1: This was only used to provide a detailed description of the owner badge
```
 .metadata(
    "description",
    format!(
        "Owner badge for resource {}",
        bech32_encoder.encode_resource_address_to_string(&resource_address)
    ),
)
```

Reply:
although arguably "the engine should be abstracted from Bech32 + NetworkDefinition" still holds as the NetworkDefinition is provided as a parameter

Reply:
It looks like it was only added a parameter for the purpose of Bech32?

Reply:
yes

Reply:
@talekhinezh I've removed it